Cannot build mysql2 on Centos 6.6 with gcc 4.5.x

Here's the output:

	Building native extensions.  This could take a while...
	ERROR:  While executing gem ... (Gem::Ext::BuildError)
		ERROR: Failed to build gem native extension.

		/usr/local/bin/ruby -r ./siteconf20150813-8428-du59mx.rb extconf.rb
	checking for ruby/thread.h... yes
	checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
	checking for rb_thread_blocking_region()... no
	checking for rb_wait_for_single_fd()... yes
	checking for rb_hash_dup()... yes
	checking for rb_intern3()... yes
	-----
	Using mysql_config at /usr/local/bin/mysql_config
	-----
	checking for mysql.h... yes
	checking for errmsg.h... yes
	checking for mysqld_error.h... yes
	-----
	Setting libpath to /usr/local/lib/mysql
	-----
	creating Makefile

	make "DESTDIR=" clean

	make "DESTDIR="
	compiling client.c
	client.c: In function ‘invalidate_fd’:
	client.c:187:43: error: ‘O_CLOEXEC’ undeclared (first use in this function)
	client.c:187:43: note: each undeclared identifier is reported only once for each function it appears in
	make: *** [client.o] Error 1

	make failed, exit code 2

	Gem files will remain installed in /usr/local/lib/ruby/gems/2.2.0/gems/mysql2-0.3.19 for inspection.
	Results logged to /usr/local/lib/ruby/gems/2.2.0/extensions/x86_64-linux/2.2.0/mysql2-0.3.19/gem_make.out

BTW, I tried both gcc 4.5.2 and 4.5.4. Also, this is mysql version if it helps:
mysql  Ver 14.14 Distrib 5.1.56, for unknown-linux-gnu (x86_64) using readline 5.1

--- a/ext/mysql2/client.c
+++ b/ext/mysql2/client.c
@@ -182,23 +182,31 @@ static void *nogvl_connect(void *ptr) {
  */
 static VALUE invalidate_fd(int clientfd)
 {
-#ifdef SOCK_CLOEXEC
+#ifdef O_CLOEXEC
   /* Atomically set CLOEXEC on the new FD in case another thread forks */
   int sockfd = open("/dev/null", O_RDWR | O_CLOEXEC);
-  if (sockfd < 0) {
-    /* Maybe SOCK_CLOEXEC is defined but not available on this kernel */
-    int sockfd = open("/dev/null", O_RDWR);
-    fcntl(sockfd, F_SETFD, FD_CLOEXEC);
-  }
 #else
-  /* Well we don't have SOCK_CLOEXEC, so just set FD_CLOEXEC quickly */
-  int sockfd = open("/dev/null", O_RDWR);
-  fcntl(sockfd, F_SETFD, FD_CLOEXEC);
+  /* Well we don't have O_CLOEXEC, trigger the fallback code below */
+  int sockfd = -1;
 #endif
 
   if (sockfd < 0) {
-    /*
-     * Cannot raise here, because one or both of the following may be true:
+    /* Either O_CLOEXEC wasn't defined at compile time, or it was defined at
+     * compile time, but isn't available at run-time. So we'll just be quick
+     * about setting FD_CLOEXEC now.
+     */
+    int flags;
+    sockfd = open("/dev/null", O_RDWR);
+    flags = fcntl(sockfd, F_GETFD);
+    /* Do the flags dance in case there are more defined flags in the future */
+    if (flags != -1) {
+      flags |= FD_CLOEXEC;
+      fcntl(sockfd, F_SETFD, flags);
+    }
+  }
+
+  if (sockfd < 0) {
+    /* Cannot raise here, because one or both of the following may be true:
      * a) we have no GVL (in C Ruby)
      * b) are running as a GC finalizer
      */

